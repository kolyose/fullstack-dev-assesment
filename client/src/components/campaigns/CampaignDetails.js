import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getCampaignById } from './../../store/actions';

import Spinner from './../spinner';

class CampaignDetails extends Component {
  componentDidMount() {
    const { campaign, match } = this.props;
    if (!campaign) {
      this.props.getCampaignById(match.params.id);
    }
  }

  render() {
    const { campaign } = this.props;
    return (
      <section>
      <div>
        <h1> Details </h1>        
        { campaign ? <p>id: { campaign.id }</p> : <Spinner/> }
      </div>
    </section>    
    );
  }
}

const mapStateToProps = ({ campaigns }, { match }) => ({
  campaign: campaigns.list.find(({ id }) => id === Number(match.params.id))
});

export default connect(mapStateToProps, { getCampaignById })(CampaignDetails);