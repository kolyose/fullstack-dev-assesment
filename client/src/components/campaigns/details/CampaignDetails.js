import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getCampaignById } from './../../../store/actions';
import config from './../../../config'

import Spinner from './../../spinner';

import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Hidden from '@material-ui/core/Hidden';
import Divider from '@material-ui/core/Divider';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import { renderStatusIcon, renderPlatformIcon } from './../../helpers';
import styles from './styles';

const renderShortDate = milliseconds => {
  const date = new Date(milliseconds);
  return `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()}`;
}

class CampaignDetails extends Component {
  componentDidMount() {
    const { campaign, match } = this.props;
    if (!campaign) {
      this.props.getCampaignById(match.params.id);
    }
  }

  render() {
    const { campaign, classes } = this.props;
    return (
      <section className={ classes.section }>
        { !campaign ? 
          <Spinner/> :
          <Card className={ classes.root }>           
            <CardHeader title={
                <div className={ classes.header }>
                  <div className={ classes.titles }>
                    <Typography variant="h5" align="left">{ campaign.name }</Typography>
                    <Typography variant="h6" align="left" color="textSecondary" noWrap>{ campaign.goal }</Typography>
                  </div>
                  <div className={ classes.status }>       
                    <Hidden xsDown>
                      <Typography variant="subtitle1" className={ classes[`status${campaign.status}`] }>
                        { campaign.status }
                      </Typography>
                    </Hidden>
                    { renderStatusIcon(campaign.status, classes) }         
                  </div>
                  <div className={ classes.budget }>
                    <Hidden xsDown>
                      <Typography variant="subtitle1">Total budget: </Typography>
                    </Hidden>
                    <Typography variant="subtitle1" className={ classes.price }>€{ campaign.total_budget }</Typography>
                  </div>
                </div>
              }/>  
            <Divider light/>
            <CardContent>
              {
               Object.entries(campaign.platforms).map(([ platformName, platformData ]) => (
                <ExpansionPanel key={ platformName }>
                  <ExpansionPanelSummary className={ classes.platformSummary } expandIcon={ <ExpandMoreIcon/> }>
                    <figure className={ classes.platformIcon }>
                      <img alt={ platformName } 
                        src={ renderPlatformIcon(platformName) }/>                  
                    </figure>    
                    <Hidden xsDown>               
                      <div className={ classes.platformName }>
                        <Typography variant="subtitle1" align="left">{ platformName }</Typography>
                      </div>
                    </Hidden>
                    <div className={ classes.platformStatus }>
                      <Hidden xsDown>
                        <Typography variant="subtitle2" className={ classes[`status${platformData.status}`] } paragraph>
                          { platformData.status }
                        </Typography>
                      </Hidden>
                      { renderStatusIcon(platformData.status, classes) }
                    </div>
                    <div className={ classes.platformBudget }>
                      <Typography variant="subtitle2" className={ classes.price }>
                        €{ `${platformData.remaining_budget}/${platformData.total_budget} `}
                      </Typography>
                    </div>
                    <div className={ classes.platformDates }>
                      <Hidden xsDown>
                        <Typography variant="subtitle2">
                          { `${new Date(platformData.start_date).toDateString()} - ${new Date(platformData.end_date).toDateString()} `}
                        </Typography>
                      </Hidden>
                      <Hidden smUp>
                        <Typography variant="subtitle2">
                          { `${renderShortDate(platformData.start_date)} - ${renderShortDate(platformData.end_date)} `}
                        </Typography>
                      </Hidden>
                    </div>
                  </ExpansionPanelSummary>
                  <Divider light/>
                  <ExpansionPanelDetails className={ classes.platformDetails }>
                    <div className={ classes.platformImage } style={
                      { 
                        width: 'auto',
                        height: 200,
                        backgroundSize: 'contain',
                        backgroundRepeat: 'no-repeat',
                        backgroundImage: `url(${config.BASE_URL}images/${platformData.creatives.image})`
                      }}
                    />                             
                    <div className={ classes.platformCreatives }>
                      <Typography variant="subtitle1">{ platformData.creatives.header }</Typography>
                      <Typography variant="subtitle2" color="textSecondary">
                        { platformData.creatives.description }
                      </Typography>
                    </div>
                  </ExpansionPanelDetails>                  
                </ExpansionPanel>
                ))
              }             
            </CardContent>            
          </Card>
        }       
      </section>
    );
  }
}

CampaignDetails.propTypes = {
  match: PropTypes.object.isRequired,
  campaign: PropTypes.shape({
    id: PropTypes.number,
    status: PropTypes.string,
    name: PropTypes.string,
    goal: PropTypes.string,
    total_budget: PropTypes.number,
    platforms:  PropTypes.object,
  }),
}

const mapStateToProps = ({ campaigns }, { match }) => ({
  campaign: campaigns.list.find(({ id }) => id === Number(match.params.id))
});

export default connect(mapStateToProps, { getCampaignById })(withStyles(styles)(CampaignDetails));