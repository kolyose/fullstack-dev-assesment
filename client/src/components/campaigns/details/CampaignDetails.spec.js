import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import data from '../../../../../server/data/data.json'
import { CampaignDetails } from './CampaignDetails.js';
import Spinner from './../../spinner';

jest.mock('./../../../store/actions')

describe('<CampaignsDetails/>', () => {

  const campaignId = 100000001;  
  let wrapper;
  let getCampaignById;

  beforeEach(() => {
    getCampaignById = jest.fn();
  })

  describe('render()', () => {
    test('renders the component', () => {
      wrapper = shallow(<CampaignDetails
        campaign={ data[0] }
        classes={{}}
        match={{ params: { id: campaignId }}}
        getCampaignById={ getCampaignById }/>);

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test("renders no <Spinner/> if 'campaign' prop is defined", () => {
      wrapper = shallow(<CampaignDetails        
        campaign={ data[0] }
        classes={{}}
        match={{ params: { id: campaignId }}}
        getCampaignById={ getCampaignById }/>);

        expect(wrapper.find(Spinner)).toHaveLength(0);
    });

    test("renders <Spinner/> if 'campaign' prop is undefined", () => {
      wrapper = shallow(<CampaignDetails        
        campaign={ undefined }
        classes={{ root: {} }}
        match={{ params: { id: campaignId }}}
        getCampaignById={ getCampaignById }/>);

        expect(wrapper.find(Spinner)).toHaveLength(1);
    });
  });

  describe('callbacks', () => {
    test("invokes 'getCampaignById()' with the campaign's id as an argument if 'campaign' prop is undefined", async () => {
      wrapper = shallow(<CampaignDetails       
        campaign={ undefined }
        classes={{}}
        match={{ params: { id: campaignId }}} 
        getCampaignById={ getCampaignById }/>);

        expect(getCampaignById).toBeCalledWith(campaignId);       
    });   
    
    test("does not invoke 'getCampaignById()' if 'campaign' prop is defined", () => {
      wrapper = shallow(<CampaignDetails
        campaign={ data[0] }
        classes={{}}
        match={{ params: { id: campaignId }}}
        getCampaignById={ getCampaignById }/>);

      expect(getCampaignById).not.toBeCalled();
    });    
  });
});