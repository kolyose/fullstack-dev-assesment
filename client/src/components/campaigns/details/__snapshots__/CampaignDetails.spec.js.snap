// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CampaignsDetails/> render() renders the component 1`] = `
<section>
  <WithStyles(Card)>
    <WithStyles(CardHeader)
      title={
        <div>
          <div>
            <WithStyles(Typography)
              align="left"
              variant="h5"
            >
              Test Ad 1
            </WithStyles(Typography)>
            <WithStyles(Typography)
              align="left"
              color="textSecondary"
              noWrap={true}
              variant="h6"
            >
              Increase Reach
            </WithStyles(Typography)>
          </div>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle1"
              >
                Delivering
              </WithStyles(Typography)>
            </Hidden>
            <pure(UpdateIcon) />
          </div>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle1"
              >
                Total budget: 
              </WithStyles(Typography)>
            </Hidden>
            <WithStyles(Typography)
              variant="subtitle1"
            >
              €
              120
            </WithStyles(Typography)>
          </div>
        </div>
      }
    />
    <WithStyles(Divider)
      light={true}
    />
    <WithStyles(CardContent)>
      <WithStyles(ExpansionPanel)
        key="facebook"
      >
        <WithStyles(ExpansionPanelSummary)
          expandIcon={<pure(ExpandMoreIcon) />}
        >
          <figure>
            <img
              alt="facebook"
              src="facebook.svg"
            />
          </figure>
          <Hidden
            implementation="js"
            lgDown={false}
            lgUp={false}
            mdDown={false}
            mdUp={false}
            smDown={false}
            smUp={false}
            xlDown={false}
            xlUp={false}
            xsDown={true}
            xsUp={false}
          >
            <div>
              <WithStyles(Typography)
                align="left"
                variant="subtitle1"
              >
                facebook
              </WithStyles(Typography)>
            </div>
          </Hidden>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                paragraph={true}
                variant="subtitle2"
              >
                Delivering
              </WithStyles(Typography)>
            </Hidden>
            <pure(UpdateIcon) />
          </div>
          <div>
            <WithStyles(Typography)
              variant="subtitle2"
            >
              €
              12/40 
            </WithStyles(Typography)>
          </div>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle2"
              >
                Mon Jul 02 2018 - Sun Jul 29 2018 
              </WithStyles(Typography)>
            </Hidden>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={true}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle2"
              >
                2018/7/2 - 2018/7/29 
              </WithStyles(Typography)>
            </Hidden>
          </div>
        </WithStyles(ExpansionPanelSummary)>
        <WithStyles(Divider)
          light={true}
        />
        <WithStyles(ExpansionPanelDetails)>
          <div
            style={
              Object {
                "backgroundImage": "url(http://localhost:8000/images/img1.jpg)",
                "backgroundRepeat": "no-repeat",
                "backgroundSize": "contain",
                "height": 200,
                "width": "auto",
              }
            }
          />
          <div>
            <WithStyles(Typography)
              variant="subtitle1"
            >
              DevOps Made Easy, We Take care of the heavy lifting for you
            </WithStyles(Typography)>
            <WithStyles(Typography)
              color="textSecondary"
              variant="subtitle2"
            >
              DOP SuperHero is where all DevOps is going to happen in the future, join the revolution today!
            </WithStyles(Typography)>
          </div>
        </WithStyles(ExpansionPanelDetails)>
      </WithStyles(ExpansionPanel)>
      <WithStyles(ExpansionPanel)
        key="instagram"
      >
        <WithStyles(ExpansionPanelSummary)
          expandIcon={<pure(ExpandMoreIcon) />}
        >
          <figure>
            <img
              alt="instagram"
              src="instagram.svg"
            />
          </figure>
          <Hidden
            implementation="js"
            lgDown={false}
            lgUp={false}
            mdDown={false}
            mdUp={false}
            smDown={false}
            smUp={false}
            xlDown={false}
            xlUp={false}
            xsDown={true}
            xsUp={false}
          >
            <div>
              <WithStyles(Typography)
                align="left"
                variant="subtitle1"
              >
                instagram
              </WithStyles(Typography)>
            </div>
          </Hidden>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                paragraph={true}
                variant="subtitle2"
              >
                Ended
              </WithStyles(Typography)>
            </Hidden>
            <pure(CheckCircleIcon) />
          </div>
          <div>
            <WithStyles(Typography)
              variant="subtitle2"
            >
              €
              14/44 
            </WithStyles(Typography)>
          </div>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle2"
              >
                Mon Jul 02 2018 - Sun Jul 29 2018 
              </WithStyles(Typography)>
            </Hidden>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={true}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle2"
              >
                2018/7/2 - 2018/7/29 
              </WithStyles(Typography)>
            </Hidden>
          </div>
        </WithStyles(ExpansionPanelSummary)>
        <WithStyles(Divider)
          light={true}
        />
        <WithStyles(ExpansionPanelDetails)>
          <div
            style={
              Object {
                "backgroundImage": "url(http://localhost:8000/images/img1.jpg)",
                "backgroundRepeat": "no-repeat",
                "backgroundSize": "contain",
                "height": 200,
                "width": "auto",
              }
            }
          />
          <div>
            <WithStyles(Typography)
              variant="subtitle1"
            >
              DevOps Made Easy, We Take care of the heavy lifting for you
            </WithStyles(Typography)>
            <WithStyles(Typography)
              color="textSecondary"
              variant="subtitle2"
            >
              DOP SuperHero is where all DevOps is going to happen in the future, join the revolution today!
            </WithStyles(Typography)>
          </div>
        </WithStyles(ExpansionPanelDetails)>
      </WithStyles(ExpansionPanel)>
      <WithStyles(ExpansionPanel)
        key="google"
      >
        <WithStyles(ExpansionPanelSummary)
          expandIcon={<pure(ExpandMoreIcon) />}
        >
          <figure>
            <img
              alt="google"
              src="google.svg"
            />
          </figure>
          <Hidden
            implementation="js"
            lgDown={false}
            lgUp={false}
            mdDown={false}
            mdUp={false}
            smDown={false}
            smUp={false}
            xlDown={false}
            xlUp={false}
            xsDown={true}
            xsUp={false}
          >
            <div>
              <WithStyles(Typography)
                align="left"
                variant="subtitle1"
              >
                google
              </WithStyles(Typography)>
            </div>
          </Hidden>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                paragraph={true}
                variant="subtitle2"
              >
                Scheduled
              </WithStyles(Typography)>
            </Hidden>
            <pure(ScheduleIcon) />
          </div>
          <div>
            <WithStyles(Typography)
              variant="subtitle2"
            >
              €
              24/36 
            </WithStyles(Typography)>
          </div>
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={true}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle2"
              >
                Mon Jul 02 2018 - Sun Jul 29 2018 
              </WithStyles(Typography)>
            </Hidden>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={false}
              smUp={true}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <WithStyles(Typography)
                variant="subtitle2"
              >
                2018/7/2 - 2018/7/29 
              </WithStyles(Typography)>
            </Hidden>
          </div>
        </WithStyles(ExpansionPanelSummary)>
        <WithStyles(Divider)
          light={true}
        />
        <WithStyles(ExpansionPanelDetails)>
          <div
            style={
              Object {
                "backgroundImage": "url(http://localhost:8000/images/img2.jpg)",
                "backgroundRepeat": "no-repeat",
                "backgroundSize": "contain",
                "height": 200,
                "width": "auto",
              }
            }
          />
          <div>
            <WithStyles(Typography)
              variant="subtitle1"
            />
            <WithStyles(Typography)
              color="textSecondary"
              variant="subtitle2"
            >
              DOP SuperHero is where all DevOps is going to happen in the future, join the revolution today!
            </WithStyles(Typography)>
          </div>
        </WithStyles(ExpansionPanelDetails)>
      </WithStyles(ExpansionPanel)>
    </WithStyles(CardContent)>
  </WithStyles(Card)>
</section>
`;
